### PR TITLE
[Sparse] Add a LADIES example

### PR DESCRIPTION
**Pull request by xiangyuzhi**
## Description
A LADIES model example use sparse library.

LADIES Algorithm:
image

Run example:
```
$ python examples/sparse/sampling/ladies.py
Training in gpu mode.
Loading data
[W TensorAdvancedIndexing.cpp:1615] Warning: scatter_reduce() is in beta and the API may change at any time. (function operator())
Training...
Epoch 00000 | Loss 2.8380 | Accuracy 0.5665 
Epoch 00001 | Loss 1.5582 | Accuracy 0.6852 
Epoch 00002 | Loss 1.2412 | Accuracy 0.7335 
Epoch 00003 | Loss 1.0939 | Accuracy 0.7564 
Epoch 00004 | Loss 1.0059 | Accuracy 0.7731 
Epoch 00005 | Loss 0.9572 | Accuracy 0.7796 
Epoch 00006 | Loss 0.9081 | Accuracy 0.7855 
Epoch 00007 | Loss 0.8707 | Accuracy 0.7935 
Epoch 00008 | Loss 0.8386 | Accuracy 0.7984 
Epoch 00009 | Loss 0.8159 | Accuracy 0.7994 
Epoch 00010 | Loss 0.7983 | Accuracy 0.8045 
Epoch 00011 | Loss 0.7795 | Accuracy 0.8060 
Epoch 00012 | Loss 0.7601 | Accuracy 0.8102 
Epoch 00013 | Loss 0.7465 | Accuracy 0.8119 
Epoch 00014 | Loss 0.7405 | Accuracy 0.8159 
Epoch 00015 | Loss 0.7234 | Accuracy 0.8197 
Epoch 00016 | Loss 0.7132 | Accuracy 0.8161 
Epoch 00017 | Loss 0.7117 | Accuracy 0.8223 
Epoch 00018 | Loss 0.6965 | Accuracy 0.8217 
Epoch 00019 | Loss 0.6871 | Accuracy 0.8237 
Testing...
Test accuracy 0.6383
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

### CODE REVIEW

**Comment by xiangyuzhi** [comment]
@dgl-bot

File: examples/sparse/sampling/ladies.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torchmetrics.functional as MF
+from dgl.data import AsNodePredDataset
+from dgl.sparse import sp_broadcast_v
+from ogb.nodeproppred import DglNodePropPredDataset
+
+
+class LADIESConv(nn.Module):
```
Comment (czkkkkkk) [inline examples/sparse/sampling/ladies.py:39]: Is this SAGEConv?

**Review by czkkkkkk — APPROVED** [review]

**Comment by czkkkkkk** [comment]
@xiangyuzhi Please also include a figure of the pseudocode for LADIES in the description.

File: examples/sparse/sampling/ladies.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+
+class SAGEConv(nn.Module):
+    r"""LADIES layer from `Layer-Dependent Importance Sampling
+    for Training Deep and Large Graph Convolutional Networks
+    <https://arxiv.org/abs/1911.07323.pdf>`__"""
+
+    def __init__(
+        self,
+        in_feats,
```
Comment (frozenbugs) [inline examples/sparse/sampling/ladies.py:46]: We prefer in_size and out_size.

Feel free to create a separate PR to change https://github.com/dmlc/dgl/blob/master/examples/sparse/sampling/graphsage.py.

File: examples/sparse/sampling/ladies.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        # Layer-wise sample nodes.
+        row_ids = torch.multinomial(row_probs, fanout, replacement=False)
+        # Add self-loop.
+        row_ids = torch.cat((row_ids, src), 0).unique()
+        sampled_matrix = sub_A.index_select(0, row_ids)
+        # Normalize edge weights.
+        div_matirx = sp_broadcast_v(
+            sampled_matrix, row_probs[row_ids].reshape(-1, 1), "truediv"
+        )
+        div_matirx = sp_broadcast_v(div_matirx, div_matirx.sum(0), "truediv")
```
Comment (frozenbugs) [inline examples/sparse/sampling/ladies.py:130]: Should this be done in sampling stage or better to do it in training stage and just return the subgraph? Either way is fine just double check.

**Review by frozenbugs — APPROVED** [review]

File: examples/sparse/sampling/ladies.py (hunk 3 above)
Comment (xiangyuzhi) [reply examples/sparse/sampling/ladies.py:130]: It chould and can only be done in sampling stage. All of these code are used to generate the subgraphs.

File: examples/sparse/sampling/ladies.py (hunk 2 above)
Comment (xiangyuzhi) [reply examples/sparse/sampling/ladies.py]: OK, PR is here https://github.com/dmlc/dgl/pull/6581

**Pull request merged**